Update ERC-8126: Move to Review (#1608)
ethereum/ERCs: ERCS/erc-8126.md

Reviewer: This reads like the reasoning behind your design (as opposed to introducing requirements), and so it belongs in the Rationale section.

Reviewer: Average -> mean?

| **OWASP** | Open Worldwide Application Security Project - nonprofit organization providing security standards and testing guides for web and smart contract applications |
| **PDV** | Private Data Verification - generates ZKPs from verification results |
| **Proof ID** | A unique identifier for a ZKP generated during verification |
| **QCV** | Quantum Cryptography Verification - provides quantum-resistant encryption for sensitive data |
| **Risk Score** | A numerical value from 0-100 indicating the assessed risk level, where 0 is the lowest risk, and 100 is the highest risk |
| **SCV** | Solidity Code Verification - validates Solidity Code security |
| **Verification Provider** | A service implementing this standard's verification types (ETV, MCV, PDV, QCV, SCV, WAV, WV) |
| **WAV** | Web Application Verification - checks endpoint security and accessibility |
| **WV** | Wallet Verification - assesses wallet history and threat database status |
| **ZKP** | Zero-Knowledge Proof - cryptographic proof that verification occurred without revealing underlying data |

## Specification

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt) and [RFC 8174](https://www.ietf.org/rfc/rfc8174.txt).

Reviewer 2: ```suggestion
The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119) and [RFC 8174](https://www.rfc-editor.org/rfc/rfc8174).
```

---
eip: 8126
title: AI Agent Verification
description: Specialized multi-layer verification for AI Agent security
author: Leigh Cronian (@cybercentry) <leigh.cronian@cybercentry.co.uk>
discussions-to: https://ethereum-magicians.org/t/erc-8126-ai-agent-registration-and-verification/27445
status: Review
type: Standards Track
category: ERC
created: 2026-01-15
requires: 155, 191, 712, 721, 3009, 8004
---

## Abstract

> This ERC defines a standard interface for verifying AI agents on Ethereum that have been registered via [ERC-8004](./eip-8004.md). It enables AI agents to undergo several specialized verification processes defined in this proposal: Ethereum Token Verification (ETV), Staking Contract Verification (SCV), Web Application Verification (WAV), and Wallet Verification (WV). Verification providers implement this standard using Private Data Verification (PDV) to generate Zero-Knowledge Proofs (ZKPs). Detailed verification results are accessible only to the AI agent’s wallet holder and include a unified risk score (0-100) to help users assess the agent’s trustworthiness. Verification attestations MAY be posted to ERC-8004’s Validation Registry for ecosystem-wide discoverability.

Reviewer: ```suggestion
This ERC defines a standard interface for verifying AI agents on Ethereum that have been registered via [ERC-8004](./eip-8004.md). It enables AI agents to undergo several specialized verification processes defined in this proposal: Ethereum Token Verification (ETV), Staking Contract Verification (SCV), Web Application Verification (WAV), and Wallet Verification (WV). Verification providers implement this standard using Private Data Verification (PDV) to generate Zero-Knowledge Proofs (ZKPs). Detailed verification results are accessible only to the AI agent’s wallet holder and include a unified risk score (0-100) to help users assess the agent’s trustworthiness. Verification attestations can additionally be posted to ERC-8004’s Validation Registry for ecosystem-wide discoverability.
```

You should avoid creating requirements outside of the specification section. They're easy to miss, and it's a bit weird to have a requirement in the abstract.

- MUST check contract against known vulnerability patterns
- MUST produce a risk score between 0 and 100
- SHOULD follow OWASP Smart Contract Security guidelines

#### Staking Contract Verification (SCV)
Validates the legitimacy and security of the staking smart contract when a `stakingContractAddress` is present in the resolved metadata.

- MUST verify that the `stakingContractAddress` is a deployed smart contract on the resolved `chain_id` by calling eth_getCode and confirming the returned bytecode is not 0x
- MUST check staking mechanism implementation
- MUST verify reward distribution logic
- MUST check for common staking vulnerabilities (reentrancy, flash loan attacks)
- MUST produce a risk score between 0 and 100
- SHOULD follow OWASP Smart Contract Security guidelines

#### Web Application Verification (WAV)
Ensures the agent's web endpoint is accessible and secure using resolved metadata.

- MUST verify HTTPS endpoint responds (using resolved `url` or `endpoints` array)
- MUST check for common security vulnerabilities
- MUST verify SSL certificate validity
- MUST produce a risk score between 0 and 100
- SHOULD follow OWASP Web Application Security Testing guidelines (WSTG)

> - SHOULD check for OWASP Top 10 vulnerabilities

Author: Thank you Sam, I've attached my Call for Input proposal. 
[20260407-Call_for_Input_ERC_8126-LSC.md](https://github.com/user-attachments/files/26523778/20260407-Call_for_Input_ERC_8126-LSC.md)

         |                +--------+----------+
         |                |       WAV         |
         |                |  (always runs)    |
         |                +--------+----------+
         |                         |
         |                         v
         |                +--------+----------+
         |                |       WV          |
         |                |  (always runs)    |
         |                +--------+----------+
         |                         |
         | 2. Aggregate Results    |
         |<------------------------+
         |
         | 3. Generate ZK Proofs via PDV
         v
+-------------------+
|       PDV         |
| (Zero-Knowledge   |
|  Proof Generation)|
+--------+----------+
         |
         | 4. Optional: QCV encryption
         v
+-------------------+
|       QCV         |
| (Quantum-Resistant|
|    Encryption)    |
+--------+----------+
         |
         | 5. Return Proof IDs + Risk Score
         v
(Optional) Post to the ERC-8004 Validation Registry
```

### Verification Types

Verification requests MUST reference an ERC-8004 `agentId` (uint256 token ID from the Identity Registry).  

> The provider MUST call `tokenURI(agentId)` on the canonical Identity Registry contract and resolve the returned URI to fetch the agent registration JSON file. All metadata fields (walletAddress/agentWallet, url/endpoints, contractAddress, stakingContractAddress, chainId, platformId, etc.) MUST be extracted from this JSON in accordance with the ERC-8004 registration schema.  

Reviewer: ```suggestion
The provider MUST call `tokenURI(agentId)` on the canonical Identity Registry contract and resolve the returned URI to fetch the agent registration JSON file. All metadata fields (`walletAddress`/`agentWallet`, `url`/`endpoints`, `contractAddress`, `stakingContractAddress`, `chainId`, `platformId`, etc.) MUST be extracted from this JSON in accordance with the ERC-8004 registration schema.  
```

I'm assuming these are specific keys/identifiers, and so should be in backticks.

Direct submission of individual parameters (wallet_address, url, contract_address, etc.) without an `agentId` is NOT permitted.

Compliant verification providers MUST implement **all** of the following verification types and apply them using the resolved metadata:

#### ETV (Ethereum Token Verification)
Validates on-chain presence and smart contract legitimacy using resolved metadata.

- MUST verify contract exists on the resolved chain_id (or mainnet if unspecified)
- MUST check contract against known vulnerability patterns
- MUST return risk score 0-100
- SHOULD follow OWASP Smart Contract Security guidelines

> #### SCV (Staking Contract Verification)

Reviewer: ```suggestion
#### Staking Contract Verification (SCV)
```

Pre-existing implementations that relied on the original custom registration logic in this ERC would need to migrate by:

- Minting agents in the ERC-8004 Identity Registry
- Setting the tokenURI to a compatible metadata JSON file containing name, description, walletAddress, url, contractAddress, stakingContractAddress, etc.

Existing AI agents already registered via ERC-8004 can undergo verification without changes. The standard is designed to work alongside existing token standards ([ERC-20](./eip-20.md), [ERC-721](./eip-721.md), [ERC-1155](./eip-1155.md)) and identity standards.

## Test Cases

### Verification Tests

- MUST complete ETV when `contractAddress` is present in resolved metadata
- MUST complete WV for all agentIds
- MUST complete WAV for all agentIds
- MUST complete SCV when `stakingContractAddress` is present in resolved metadata
- MUST generate PDV ZKPs for each verification type
- MUST calculate overallRiskScore as the average of applicable scores
- MUST emit AgentVerified event with all proof IDs (if on-chain component is deployed)
- MUST revert with `VerificationFailed` on provider error

> - MUST revert with `InsufficientCredits` when no credits are available

Reviewer: Same comment about requirements outside of the Specification section.
